File: portara/src/main/java/io/kaleido/kata/Request.java
/*
 * Copyright © 2024 Kaleido, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package io.kaleido.kata;

import java.util.UUID;

import github.com.kaleido_io.paladin.kata.Kata;

// Create a class that includes a method receiving a callback function
public abstract class Request {
    private Handler transactionHandler;

    private ResponseHandler responseHandler;
    private final String id;

    public Request(Handler transactionHandler, ResponseHandler responseHandler) {
        this.transactionHandler = transactionHandler;
        this.responseHandler = responseHandler;
                this.id = UUID.randomUUID().toString();

    }

    public String getId() {
        return this.id;
    }

    // Method that receives a callback function
    public void send(ResponseHandler responseHandler) throws Exception {
        this.transactionHandler.submitTransaction(this);
    }
    
    public abstract Kata.Message getRequestMessage();

    public ResponseHandler getResponseHandler() {
        return responseHandler;
    }

    public Handler getTransactionHandler() {
        return transactionHandler;
    }
    
}

File: core/go/internal/sequencer/common/request.go
/*
 * Copyright © 2025 Kaleido, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package common

import (
	"context"

	"time"

	"github.com/google/uuid"
)

type IdempotentRequest struct {
	idempotencyKey   uuid.UUID                                                 //unique string to identify the request (re-used across retries)
	requestTime      *time.Time                                                //time the request was most recently sent
	firstRequestTime *time.Time                                                //time the request was first sent
	send             func(ctx context.Context, idempotencyKey uuid.UUID) error // function to send the request
	clock            Clock
	timeout          time.Duration
}

func NewIdempotentRequest(ctx context.Context, clock Clock, timeout time.Duration, sendRequest func(ctx context.Context, idempotencyKey uuid.UUID) error) *IdempotentRequest {
	idempotencyKey := uuid.New()
	r := &IdempotentRequest{
		idempotencyKey:   idempotencyKey,
		clock:            clock,
		send:             sendRequest,
		timeout:          timeout,
		requestTime:      nil,
		firstRequestTime: nil,
	}

	return r
}

func (r *IdempotentRequest) IdempotencyKey() uuid.UUID {
	return r.idempotencyKey
}

func (r *IdempotentRequest) FirstRequestTime() *time.Time {
	return r.firstRequestTime
}

// Prompt to check whether a retry is due and if so, send the request
func (r *IdempotentRequest) Nudge(ctx context.Context) error {
	if r.requestTime == nil || r.clock.HasExpired(*r.requestTime, r.timeout) {
		err := r.send(ctx, r.idempotencyKey)
		if err == nil {
			now := r.clock.Now()
			r.requestTime = &now
			if r.firstRequestTime == nil {
				r.firstRequestTime = r.requestTime
			}
		}
		return err
	}
	return nil
}
